Task3: Spring Boot+Liquibase: after release and  previous comments

PavelBUTORIN opened the pull request:
Added UserService class for business logic;
Added Lombok annotations;
Getters, setters and constructors were removed;
fixed number of api version: v2 -> v1;
@Autowired annotation was removed, and UserRepository field redefined as final;

PavelBUTORIN committed:
Task3: Added UserService class for business logic;
Added Lombok annotations;
Getters, setters and constructors were removed;
fixed number of api version: v2 -> v1;
@Autowired annotation was removed, and UserRepository field redefined as final;

kurortnik00 commented on pom.xml:
```diff
@@ -49,11 +50,11 @@
 			<groupId>org.projectlombok</groupId>
 			<artifactId>lombok</artifactId>
-			<optional>true</optional>
+			<version>1.18.30</version>
```
когда ты добавил <lombok.version>1.18.30</lombok.version> теперь здесь можно написать  <version>{$lombok.version}</version> тогда в случае когда тебе нужно будет поднять версию тебе нужно будет только в одном месте поправить код (а не в двух и более)

kurortnik00 replied:
чуть не так, сорян ${lombok.version}

kurortnik00 commented on pom.xml:
```diff
@@ -73,6 +74,7 @@
 						<path>
 							<groupId>org.projectlombok</groupId>
 							<artifactId>lombok</artifactId>
+							<version>1.18.30</version>
```
и вот туту <version>{$lombok.version}</version>

PavelBUTORIN committed:
Task3: Fixed little remarks
Added application.properties file;
Teachers List was relocated to UserController from TimeController;
Lombok lib version was revised and corrected;

bvsemak commented on src/main/java/ru/Pablo/TaskOne/controllers/UserController.java:
```diff
@@ -0,0 +1,32 @@
+    private final UserService userService;
+
+    @GetMapping("/users/all")
+    public List<User> allUsers() {
```
Лучше возвращать ResponseEntity, это generic обертка для http ответа, посмотри про неё обязательно, в реале голый ответы без этой обертки почти не возвращают

bvsemak replied:
allUsers не очень хорошее имя для метода, метод - это всегда какое-то действие, поэтому лучше использовать глаголы, например getAllUsers()

bvsemak commented on src/main/java/ru/Pablo/TaskOne/controllers/UserController.java:
```diff
@@ -0,0 +1,32 @@
+
+    private final UserService userService;
+
+    @GetMapping("/users/all")
```
В REST API обычно не пишут all, чтобы вернуть всех юзеров или другие сущности, а просто оставляют users, если нужен конкретный юзер, то просто делают users/{id}, ну и вообще по дефолту, если мы не указываем id сущности на какой-то запрос, то по умолчанию обычно считается, что хотим всё

bvsemak commented on src/main/java/ru/Pablo/TaskOne/controllers/UserController.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/v1")
```
Это контроллер юзеров, поэтому я бы в реквест маппинг вынес ещё users

bvsemak commented on src/main/java/ru/Pablo/TaskOne/controllers/UserController.java:
```diff
@@ -0,0 +1,32 @@
+        return userService.findAll();
+    }
+
+    @GetMapping("/teachers")
```
По сути это уже другая сущность, поэтому я бы для этого отдельный контроллер вынес. В принципе выносить отдельно что-то, что по домену отличается - хорошая практика, даже если в данный момент времени технически все похоже

bvsemak commented on src/main/java/ru/Pablo/TaskOne/controllers/UserController.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @GetMapping("/teachers")
+    public List<String> teachers(){
```
То же самое про ResponseEntity и имя метода

bvsemak commented on src/main/java/ru/Pablo/TaskOne/controllers/UserController.java:
```diff
@@ -0,0 +1,32 @@
+
+    @GetMapping("/teachers")
+    public List<String> teachers(){
+        return List.of("Ярослав Викторович", "Богдан Валерьевич", "и ещё Славик!");
```
Смотри, ты каждый раз создаешь коллекцию заново, чтобы её возвращать, для этого проекта это не страшно, но для реального приложения, а тем более если есть хоть какая-то нагрузка, то это уже порочная практика, лучше к этому не привыкать, поэтому если тебе всегда надо возвращаться одно и то же, то можно просто один раз это создать и хранить где-то, например в приватном статическом файнал поле и отдавать его по запросу, тогда у тебя не будет каждый раз тратиться время на создание коллекции. Ну и про домены, это уже какая-то логика, лучше сделать сервис для тичеров и там это реализовать

bvsemak commented on src/main/java/ru/Pablo/TaskOne/entities/User.java:
```diff
@@ -0,0 +1,37 @@
+@Entity
+@Table(name = "users")
+@Data
+public class User implements Serializable {
```
Serializable здесь не нужен, так как hibernate сам сделать все сериализации и десериализации. Вообще использовать встроенный механизм сериализации java - это устаревшая и порочная практика, так как завязывается чисто на жаву, имеет проблемы с производительностью, весом, безопасностью и кучей всего. Вместо этого используют другие механизмы, например тот же jackson

bvsemak commented on src/main/java/ru/Pablo/TaskOne/entities/User.java:
```diff
@@ -0,0 +1,37 @@
+
+@Entity
+@Table(name = "users")
+@Data
```
У аннотации @Data есть проблема, когда она ставится на Entity классе, там по-моему даже IDEA это подсвечивает, там toString и hashCode становятся ломанными

bvsemak commented on src/main/java/ru/Pablo/TaskOne/entities/User.java:
```diff
@@ -0,0 +1,37 @@
+@Data
+public class User implements Serializable {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
Для ненагруженных приложений такая стратегия пойдет, но для нагрузки уже не очень хорошо, так как он для каждого нового id будет идти в базу и просить дать ему этот id, запросы в базу сильно дольше, чем операции внутри самого приложения. Для таких случаев используют sequence и сразу у базы запрашивают условно 100 значений последовательности и выдаются уже внутри приложения, а база эти 100 никому больше не дает

bvsemak commented on src/main/java/ru/Pablo/TaskOne/services/UserService.java:
```diff
@@ -0,0 +1,18 @@
+    private final UserRepository userRepository;
+
+    public List<User> findAll() {
+        return userRepository.findAll();
```
Вопрос на подумать (сейчас не обязательно в это закапываться): а что будет, если у нас будет миллион юзеров?

bvsemak commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,15 @@
+
+spring.datasource.url=jdbc:postgresql://localhost:5432/Task3
+spring.datasource.username=PavelButorin
+spring.datasource.password=13091995
```
Тоже не обязательно все сейчас делать, но пароли в открытом виде обычно не прописывают в исходниках, а берут из других мест, например из переменных окружения, если будет время, то посмотри как передать пароль через переменные окружения и зачем это

bvsemak reviewed (commented):
Молодец! Хорошо идешь

PavelBUTORIN committed:
Task3: Fixed after Mr.Bogdan's remarks:
Added TeacherController;
Teachers List was removed from UserController;
Lombok User Class annotations have been expanded;
ResponseEntity class was used in UserController and TeacherController;
URLs were fixed;
Serializable implementation was removed from User class determination;

bvsemak commented on src/main/java/ru/Pablo/TaskOne/controllers/TeacherController.java:
```diff
@@ -0,0 +1,17 @@
+public class TeacherController {
+    @GetMapping("/teachers")
+    public ResponseEntity<List<String>> getTeachers(){
+        return ResponseEntity.ok(List.of("Ярослав Викторович", "Богдан Валерьевич", "и ещё Славик!"));
```
Извините за душность, но такое лучше выносить в сервисный слой

bvsemak commented on src/main/java/ru/Pablo/TaskOne/entities/User.java:
```diff
@@ -0,0 +1,42 @@
+@Table(name = "users")
+@Getter
+@Setter
+@EqualsAndHashCode
+@ToString
```
Вот эти две аннотации как раз и имеют проблему с Entity. Аннотация Data - это просто группирующая аннотация, которая содержит в себе все те аннотации, которые Вы тут руками внесли (можете посмотреть, что там внутри), поэтому проблема не ушла, а просто было переписано все в другом виде, как раз аннотацию Data не просто убирают и все содержимое её переносят, кроме аннотаций ToString и EqualsAndHashCode

bvsemak commented on src/main/java/ru/Pablo/TaskOne/services/UserService.java:
```diff
@@ -0,0 +1,26 @@
+        return userRepository.findAll();
+    }
+
+    public ResponseEntity<User> findById(Long id) {
```
В сервисном слое ResponseEntity обычно не возвращают, это нарушение области ответственности по сути, сервис юзеров занимается бизнес логикой юзеров и все, а ResponseEntity это уже детали работы с http протоколом, такое должно находиться в контроллере по смыслу, т.е. здесь мы возвращаем просто User, а в контроллере уже оборачиваем в ResponseEntity.

bvsemak commented on src/main/java/ru/Pablo/TaskOne/entities/User.java:
```diff
@@ -0,0 +1,42 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name="id")
+    private long id;
```
Id должен быть объектовым Long, а не примитивным, иначе Hibernate будет ломаться, потому что он работает с null id активно и принимает на этом основании много решение по персисту

bvsemak commented on src/main/java/ru/Pablo/TaskOne/services/UserService.java:
```diff
@@ -0,0 +1,26 @@
+
+    public ResponseEntity<User> findById(Long id) {
+
+        var user = userRepository.findAll().stream().filter(p->(p.getId()==id)).findFirst();
```
У jpaRepository есть метод findById, стрим здесь не нужен, тем более в случае больших объемов данных ты будешь выгружать всю базу в приложение и внутри производить фильтрацию, так лучше не делать, все-таки база данных как раз для этого и нужна, чтобы операции с данными производить в ней эффективно.
Плюс сравнение через равноравно getId() == id неправильное, когда id будет объектом, а не примитивом, так как будет сравниваться ссылка, а не значение после определенных номеров (256 малых значений закешированы, для них прокатит)

bvsemak commented on src/main/java/ru/Pablo/TaskOne/services/UserService.java:
```diff
@@ -0,0 +1,26 @@
+    public ResponseEntity<User> findById(Long id) {
+
+        var user = userRepository.findAll().stream().filter(p->(p.getId()==id)).findFirst();
+        System.out.println(user);
```
Тут нужно ознакомиться с логированием, в реале так никто не пишет для вывода информации

bvsemak reviewed (commented):
Немного поддушнил, сори

PavelBUTORIN committed:
Task3: Fixed after Mr.Bogdan's remarks 2 (a.k.a. Dormammu, I've come to bargain!):
Added TeacherService and TeacherList-part from TeacherController was moved there;
long -> Long;
Lombok User Class annotations have been updated;
ResponseEntity class was used in UserController and TeacherController in more correct way;
Password was converted to environment variables;
FindById fixed;

bvsemak reviewed (approved):
Всё гуд
